File: ios/Classes/FmodFlutterPlugin.swift
import Flutter
import UIKit

public class FmodFlutterPlugin: NSObject, FlutterPlugin {
    private var fmodManager: FmodManager?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "fmod_flutter", binaryMessenger: registrar.messenger())
        let instance = FmodFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            handleInitialize(result: result)
        case "loadBanks":
            handleLoadBanks(call: call, result: result)
        case "playEvent":
            handlePlayEvent(call: call, result: result)
        case "stopEvent":
            handleStopEvent(call: call, result: result)
        case "setParameter":
            handleSetParameter(call: call, result: result)
        case "setPaused":
            handleSetPaused(call: call, result: result)
        case "setVolume":
            handleSetVolume(call: call, result: result)
        case "update":
            fmodManager?.update()
            result(nil)
        case "release":
            fmodManager?.release()
            fmodManager = nil
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func handleInitialize(result: @escaping FlutterResult) {
        fmodManager = FmodManager()
        let success = fmodManager?.initialize() ?? false
        result(success)
    }
    
    private func handleLoadBanks(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let banks = args["banks"] as? [String] else {
            result(FlutterError(code: "INVALID_ARGS", message: "Banks list required", details: nil))
            return
        }
        
        let success = fmodManager?.loadBanks(banks) ?? false
        result(success)
    }
    
    private func handlePlayEvent(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["path"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Event path required", details: nil))
            return
        }
        
        fmodManager?.playEvent(path)
        result(nil)
    }
    
    private func handleStopEvent(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["path"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Event path required", details: nil))
            return
        }
        
        fmodManager?.stopEvent(path)
        result(nil)
    }
    
    private func handleSetParameter(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["path"] as? String,
              let parameter = args["parameter"] as? String,
              let value = args["value"] as? Double else {
            result(FlutterError(code: "INVALID_ARGS", message: "Path, parameter, and value required", details: nil))
            return
        }
        
        fmodManager?.setParameter(path: path, paramName: parameter, value: Float(value))
        result(nil)
    }
    
    private func handleSetPaused(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["path"] as? String,
              let paused = args["paused"] as? Bool else {
            result(FlutterError(code: "INVALID_ARGS", message: "Path and paused state required", details: nil))
            return
        }
        
        fmodManager?.setPaused(path: path, paused: paused)
        result(nil)
    }
    
    private func handleSetVolume(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let path = args["path"] as? String,
              let volume = args["volume"] as? Double else {
            result(FlutterError(code: "INVALID_ARGS", message: "Path and volume required", details: nil))
            return
        }
        
        fmodManager?.setVolume(path: path, volume: Float(volume))
        result(nil)
    }
}

File: android/src/main/kotlin/com/midnightlaunchgames/fmod_flutter/FmodFlutterPlugin.kt
package com.midnightlaunchgames.fmod_flutter

import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FmodFlutterPlugin */
class FmodFlutterPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel
  private lateinit var context: Context
  private lateinit var fmodManager: FmodManager

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    context = flutterPluginBinding.applicationContext
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "fmod_flutter")
    channel.setMethodCallHandler(this)
    fmodManager = FmodManager(context)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "initialize" -> {
        result.success(fmodManager.initialize())
      }
      "loadBanks" -> {
        val banks = call.argument<List<String>>("banks")
        if (banks != null) {
          result.success(fmodManager.loadBanks(banks))
        } else {
          result.error("INVALID_ARGS", "Banks list required", null)
        }
      }
      "playEvent" -> {
        val path = call.argument<String>("path")
        if (path != null) {
          fmodManager.playEvent(path)
          result.success(null)
        } else {
          result.error("INVALID_ARGS", "Event path required", null)
        }
      }
      "stopEvent" -> {
        val path = call.argument<String>("path")
        if (path != null) {
          fmodManager.stopEvent(path)
          result.success(null)
        } else {
          result.error("INVALID_ARGS", "Event path required", null)
        }
      }
      "setParameter" -> {
        val path = call.argument<String>("path")
        val param = call.argument<String>("parameter")
        val value = call.argument<Double>("value")
        if (path != null && param != null && value != null) {
          fmodManager.setParameter(path, param, value.toFloat())
          result.success(null)
        } else {
          result.error("INVALID_ARGS", "Path, parameter, and value required", null)
        }
      }
      "setPaused" -> {
        val path = call.argument<String>("path")
        val paused = call.argument<Boolean>("paused")
        if (path != null && paused != null) {
          fmodManager.setPaused(path, paused)
          result.success(null)
        } else {
          result.error("INVALID_ARGS", "Path and paused state required", null)
        }
      }
      "setMasterPaused" -> {
        val paused = call.argument<Boolean>("paused")
        if (paused != null) {
          result.success(fmodManager.setMasterPaused(paused))
        } else {
          result.error("INVALID_ARGS", "Paused state required", null)
        }
      }
      "setVolume" -> {
        val path = call.argument<String>("path")
        val volume = call.argument<Double>("volume")
        if (path != null && volume != null) {
          fmodManager.setVolume(path, volume.toFloat())
          result.success(null)
        } else {
          result.error("INVALID_ARGS", "Path and volume required", null)
        }
      }
      "update" -> {
        fmodManager.update()
        result.success(null)
      }
      "release" -> {
        fmodManager.release()
        result.success(null)
      }
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    fmodManager.release()
  }
}

